text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+Option 1 essentially reverts all transactions that happened after the block whose FTT has been exceeded. Option 2 attempts to revert the transactions from all forks but one. However, it is possible that some of the transactions on different forks of the chain don't conflict with each other. In such a case, we could attempt to replay all of them on the new trusted branch. This is the idea behind option 3: to gather the non-conflicting transactions, create a block containing them and collectively declare it correct, outside of the consensus algorithm.
+
+After the users collectively decide upon the last correct block, the transactions to be replayed and the list of equivocators to be slashed, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the last correct block, the transactions that need to be replayed on top of it and the set of equivocating validators (exchanging evidence if necessary).
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
+    - the initial set of validators for the restarted network which excludes the equivocators,
+    - the initial global state hash for the restarted network
+5. The network is being restarted using the chainspec constructed above.
```

Comment:
I think that after detecting an equivocation and (socially) coming to consensus on which block (from the Linear chain) is the last trusted one you can just use the same mechanism as for the node sync/join described in my document (CEP-0005).

Reply:
Except that we'll (in the long run) need to provide tools to make the required changes to the global state (e.g. removing validators) and use the new global state as the starting point, instead of only the post-state of the last good block.

Reply:
Hmm, I thought that the point of social consensus was to agree on a point in the history at which there's no equivocation – i.e. is the last _safe_ block. Maybe that equivocation was caused by a bug and not an intentional attack etc.

Modifying the consensus' state, like getting rid of equivocators, is a different concern.

Reply:
I'd say it's part of it. After an equivocation catastrophe, we _can't_ continue from the last good block without modification: The unmodified validator set already contains too many equivocators to continue.

Reply:
By the "last good block" I meant one from before (any, some?) equivocations. If the equivocation was a result of software bug, patching software and starting over from the last good block is enough. If starting over would lead to equivocations again, then you're right – we cannot continue with equivocators as validators.

Reply:
Mateusz makes an excellent point - there needs to be a differentiation in these procedures for handling software bugs and actual attacks, to the extent that we can distinguish them. We can't just burn millions of someone's dollars because we pushed a bad patch.